File: VulpesCloud-node/src/main/kotlin/io/github/thecguygithub/node/command/provider/CommandProviderImpl.java
package io.github.thecguygithub.node.command.provider;

import com.google.common.collect.Iterables;
import io.github.thecguygithub.api.command.CommandInfo;
import io.github.thecguygithub.node.Node;
import io.github.thecguygithub.node.command.CommandProvider;
import io.github.thecguygithub.node.command.annotations.AliasAnnotation;
import io.github.thecguygithub.node.command.annotations.Description;
import io.github.thecguygithub.node.command.defaults.DefaultCommandManager;
import io.github.thecguygithub.node.command.source.CommandSource;
import io.github.thecguygithub.node.logging.Logger;
import io.leangen.geantyref.TypeToken;
import jakarta.inject.Inject;
import jakarta.inject.Singleton;
import lombok.NonNull;
import org.incendo.cloud.CommandManager;
import org.incendo.cloud.annotations.AnnotationParser;
import org.incendo.cloud.key.CloudKey;
import org.incendo.cloud.meta.CommandMeta;
import org.jetbrains.annotations.Nullable;

import java.util.*;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;

// @Singleton
public class CommandProviderImpl implements CommandProvider {

    private static final CloudKey<Set<String>> ALIAS_KEY = CloudKey.of("testcloud:alias", new TypeToken<Set<String>>() {});
    private static final CloudKey<String> DESCRIPTION_KEY = CloudKey.of("testcloud:description", String.class);

    private final CommandManager<CommandSource> commandManager;
    private final AnnotationParser<CommandSource> annotationParser;
    private final List<CommandInfo> registeredCommands;


    // @Inject
    public CommandProviderImpl(
            @NonNull DefaultCommandManager commandManager
    ) {

        Node.Companion.getTerminal().printLine("WEPWOP");

        this.commandManager = commandManager;

        Node.Companion.getTerminal().printLine("WOPWEP");
        this.annotationParser = new AnnotationParser<>(
                this.commandManager,
                CommandSource.class,
                _ -> CommandMeta.empty()
        );

        Node.Companion.getTerminal().printLine("KRUEMEL MC");

        this.annotationParser.registerBuilderModifier(
                AliasAnnotation.class,
                (alias, builder) -> builder.meta(ALIAS_KEY, new HashSet<>(Arrays.asList(alias.value()))));
        this.annotationParser.registerBuilderModifier(Description.class, (description, builder) -> {
            if (!description.value().trim().isEmpty()) {
                return builder.meta(DESCRIPTION_KEY, description.value());
            }
            return builder;
        });

        Node.Companion.getTerminal().printLine("ALRAM!");

        this.registeredCommands = new ArrayList<>();

    };

    @Override
    public @NonNull List<String> suggest(@NonNull CommandSource source, @NonNull String input) {
        return List.of();
    }

    @Override
    public CompletableFuture<?> execute(@NonNull CommandSource source, @NonNull String input) {
        return this.commandManager.commandExecutor().executeCommand(source, input).exceptionally(exception -> {
            throw exception instanceof CompletionException cex ? cex : new CompletionException(exception);
        });

    }

    @Override
    public void register(@NonNull Class<?> commandClass) {

    }

    @Override
    public void register(@NonNull Object command) {

        var cloudCommand = Iterables.getFirst(this.annotationParser.parse(command), null);

        if (cloudCommand != null) {
            if (cloudCommand.nonFlagArguments().isEmpty()) {
                return;
            }

            var permission = cloudCommand.commandPermission().permissionString();

            var description = cloudCommand.commandMeta().get(DESCRIPTION_KEY);

            var aliases = cloudCommand.commandMeta().getOrDefault(ALIAS_KEY, Collections.emptySet());

            var name = cloudCommand.nonFlagArguments().getFirst().name().toLowerCase(Locale.ROOT);

            this.registeredCommands.add(
                    new CommandInfo(name, aliases, description, this.commandUsageOfRoot(name))
            );
        }

    }

    @Override
    public @Nullable CommandInfo command(@NonNull String name) {
        var lowerCaseInput = name.toLowerCase(Locale.ROOT);
        for (var command : this.registeredCommands) {
            if (command.name().equals(lowerCaseInput) || command.aliases().contains(lowerCaseInput)) {
                return command;
            }
        }
        return null;
    }

    @Override
    public @NonNull Collection<CommandInfo> commands() {
        return Collections.unmodifiableCollection(this.registeredCommands);
    }

    private @NonNull List<String> commandUsageOfRoot(@NonNull String root) {
        List<String> commandUsage = new ArrayList<>();
        for (var command : this.commandManager.commands()) {
            // the first argument is the root, check if it matches
            var arguments = command.components();
            if (arguments.isEmpty() || !arguments.getFirst().name().equalsIgnoreCase(root)) {
                continue;
            }

            commandUsage.add(this.commandManager.commandSyntaxFormatter().apply(null, arguments, null));
        }

        Collections.sort(commandUsage);
        return commandUsage;
    }

}

File: VulpesCloud-node/src/main/kotlin/de/vulpescloud/node/command/impl/CommandProviderImpl.kt
package de.vulpescloud.node.command.impl

import com.github.benmanes.caffeine.cache.Caffeine
import com.google.common.collect.Iterables
import de.vulpescloud.api.lang.Translator
import de.vulpescloud.node.command.CloudCommandManager
import de.vulpescloud.node.command.CommandProvider
import de.vulpescloud.node.command.CommandSource
import de.vulpescloud.node.command.annotations.Alias
import de.vulpescloud.node.command.annotations.Description
import org.incendo.cloud.Command
import org.incendo.cloud.CommandManager
import org.incendo.cloud.annotations.AnnotationParser
import org.incendo.cloud.annotations.BuilderModifier
import org.incendo.cloud.execution.CommandResult
import org.incendo.cloud.key.CloudKey
import org.incendo.cloud.meta.CommandMeta
import org.incendo.cloud.processors.cache.CaffeineCache
import org.incendo.cloud.processors.confirmation.ConfirmationConfiguration
import org.incendo.cloud.processors.confirmation.ConfirmationManager
import org.incendo.cloud.processors.confirmation.annotation.ConfirmationBuilderModifier
import org.slf4j.LoggerFactory
import java.time.Duration
import java.util.*
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionException



class CommandProviderImpl(
    private val translator: Translator,
) : CommandProvider {

    private val aliasKey: CloudKey<Array<String>>? = CloudKey.of("vulpescloud:alias", Array<String>::class.java)
    private val descriptionKey: CloudKey<String> = CloudKey.of(
        "vulpescloud:description",
        String::class.java
    )
    private var registeredCommands: MutableList<CommandInfo>? = mutableListOf()
    private var annotationParser: AnnotationParser<CommandSource>? = null

    override val commandManager: CommandManager<CommandSource> = CloudCommandManager()
    private val logger = LoggerFactory.getLogger(CommandProviderImpl::class.java)

    override fun initialize() {
        this.annotationParser = AnnotationParser(
            this.commandManager,
            CommandSource::class.java
        ) { CommandMeta.empty() }

        this.annotationParser!!.registerBuilderModifier(
            Alias::class.java,
            BuilderModifier<Alias, CommandSource?> registerBuilderModifier@{ alias: Alias, builder: Command.Builder<CommandSource?> ->
                if (alias.alias.isNotEmpty()) {
                    return@registerBuilderModifier builder.meta(
                        this.aliasKey!!,
                        alias.alias
                    )
                }
                builder
            }
        )

        this.annotationParser!!.registerBuilderModifier(
            Description::class.java,
            BuilderModifier<Description, CommandSource?> registerBuilderModifier@{ description: Description, builder: Command.Builder<CommandSource?> ->
                if (description.description.trim { it <= ' ' }.isNotEmpty()) {
                    return@registerBuilderModifier builder.meta<String>(
                        this.descriptionKey,
                        if (description.translatable) {
                            translator.trans(description.description)
                        } else {
                            description.description
                        }
                    )
                }
                builder
            }
        )

        ConfirmationBuilderModifier.install(this.annotationParser!!)
        val confirmationManager = ConfirmationManager.confirmationManager(
            ConfirmationConfiguration.builder<CommandSource>()
                .cache(CaffeineCache.of(Caffeine.newBuilder().expireAfterWrite(Duration.ofSeconds(30)).build()))
                .noPendingCommandNotifier { it!!.sendMessage("There are no Commands that need confirmation") }
                .confirmationRequiredNotifier { source, _ -> source!!.sendMessage("Type 'confirm' to confirm this command") }
                .build()
        )
        this.commandManager.registerCommandPostProcessor(confirmationManager.createPostprocessor())
        this.commandManager.command(
            this.commandManager.commandBuilder("confirm").handler(confirmationManager.createExecutionHandler())
        )

        registeredCommands!!.add(
            CommandInfo(
                "confirm",
                setOf(),
                "",
                listOf(""),
                ""
            )
        )

    }

    override fun suggest(source: CommandSource, input: String): List<String> {
        return listOf()
    }

    override fun execute(source: CommandSource, input: String): CompletableFuture<CommandResult<CommandSource>> {
        return commandManager.commandExecutor().executeCommand(source, input).exceptionally { exception: Throwable? ->
            logger.error("Exception while executing command", exception)
            throw if (exception is CompletionException) exception else CompletionException(exception)
        }
    }

    override fun register(command: Any) {
        val cloudCommand = Iterables.getFirst(
            annotationParser!!.parse(command), null
        )

        if (cloudCommand != null) {
            if (cloudCommand.nonFlagArguments().isEmpty()) {
                return
            }

            val permission = cloudCommand.commandPermission().permissionString()

            val description = cloudCommand.commandMeta().getOrSupplyDefault(
                descriptionKey
            ) { "No Description!" }

            val aliases = cloudCommand.commandMeta().getOrSupplyDefault(
                aliasKey ?: CloudKey.of("vulpescloud:alias", Array<String>::class.java)
            ) { emptyArray() }

            val name = cloudCommand.nonFlagArguments().first().name().lowercase()

            registeredCommands!!.add(
                CommandInfo(name, aliases.toSet(), description, this.commandUsageOfRoot(name), permission)
            )
        }
    }

    override fun command(name: String): CommandInfo? {
        val lowerCaseInput = name.lowercase()
        for (command in registeredCommands!!) {
            if (command.name == lowerCaseInput || command.aliases.contains(lowerCaseInput)) {
                return command
            }
        }
        return null
    }

    override fun commands(): MutableCollection<CommandInfo>? {
        return this.registeredCommands?.let { Collections.unmodifiableCollection(it) }
    }

    override fun commandUsageOfRoot(root: String): List<String> {
        val commandUsage: MutableList<String> = ArrayList()
        for (command in commandManager.commands()) {
            // the first argument is the root, check if it matches
            val arguments = command.components()
            if (arguments.isEmpty() || !arguments.first().name().equals(root, ignoreCase = true)) {
                continue
            }

            commandUsage.add(commandManager.commandSyntaxFormatter().apply(null, arguments, null))
        }

        commandUsage.sort()
        return commandUsage
    }
}